File: android-native/app/src/main/java/app/dimo/android/domain/Formatting.kt
package app.dimo.android.domain

import app.dimo.android.data.model.Currency
import java.text.NumberFormat
import java.time.LocalTime
import java.util.Locale
import kotlin.math.abs
import kotlin.math.roundToInt

/**
 * Port of `ios-native/Dimo/Domain/Formatting.swift`.
 *
 * The Swift implementation pins `NumberFormatter` to `en_IN`, which produces the
 * lakh/crore grouping (`₹1,00,000`). Keep that locale so all three clients render
 * identical strings.
 */
object Formatting {
  private val INDIAN_LOCALE: Locale = Locale.forLanguageTag("en-IN")

  private val symbols = mapOf(
    Currency.INR to "₹",
    Currency.USD to "$",
    Currency.EUR to "€",
  )

  fun currencySymbol(currency: Currency = Currency.INR): String = symbols[currency] ?: "₹"

  fun money(amount: Double, currency: Currency = Currency.INR): String =
    moneyWithSymbol(amount, currencySymbol(currency))

  /** Format money for any currency code (default or a foreign entry currency). */
  fun money(amount: Double, currencyCode: String): String =
    moneyWithSymbol(amount, CurrencyMeta.symbol(currencyCode))

  fun decimal(value: Double, maximumFractionDigits: Int): String {
    val formatter = NumberFormat.getNumberInstance(INDIAN_LOCALE)
    formatter.minimumFractionDigits = 0
    formatter.maximumFractionDigits = maximumFractionDigits
    return formatter.format(value)
  }

  private fun moneyWithSymbol(amount: Double, symbol: String): String {
    val hasFraction = abs(amount % 1) > 0.0001
    val formatter = NumberFormat.getNumberInstance(INDIAN_LOCALE)
    formatter.minimumFractionDigits = if (hasFraction) 2 else 0
    formatter.maximumFractionDigits = 2
    val formatted = formatter.format(abs(amount))
    // U+2212 MINUS SIGN, matching the Swift implementation.
    return (if (amount < 0) "−" else "") + symbol + formatted
  }

  fun spent(amount: Double, currency: Currency = Currency.INR): String =
    "−" + money(amount, currency)

  fun percent(value: Double, total: Double): Int {
    if (total <= 0) return 0
    return (value / total * 100).roundToInt()
  }

  fun compactMoney(amount: Double, currency: Currency = Currency.INR): String {
    val symbol = currencySymbol(currency)
    if (amount >= 1000) {
      val k = String.format(Locale.ROOT, "%.1f", amount / 1000).replace(".0", "")
      return "$symbol${k}k"
    }
    val trimmed = String.format(Locale.ROOT, "%.2f", amount).replace(Regex("\\.?0+$"), "")
    return "$symbol$trimmed"
  }
}

object Greeting {
  fun greetingFor(time: LocalTime = LocalTime.now(DateHelpers.zone())): String {
    val hour = time.hour
    if (hour < 12) return "Good morning"
    if (hour < 17) return "Good afternoon"
    return "Good evening"
  }
}

File: ios-native/Dimo/Domain/Formatting.swift
import Foundation

enum Formatting {
  private static let symbols: [Currency: String] = [
    .INR: "₹",
    .USD: "$",
    .EUR: "€",
  ]

  static func currencySymbol(_ currency: Currency = .INR) -> String {
    symbols[currency] ?? "₹"
  }

  static func money(_ amount: Double, currency: Currency = .INR) -> String {
    money(amount, symbol: currencySymbol(currency))
  }

  /// Format money for any currency code (default or a foreign entry currency).
  static func money(_ amount: Double, currencyCode: String) -> String {
    money(amount, symbol: CurrencyMeta.symbol(currencyCode))
  }

  static func decimal(_ value: Double, maximumFractionDigits: Int) -> String {
    let formatter = NumberFormatter()
    formatter.locale = Locale(identifier: "en_IN")
    formatter.numberStyle = .decimal
    formatter.minimumFractionDigits = 0
    formatter.maximumFractionDigits = maximumFractionDigits
    return formatter.string(from: NSNumber(value: value)) ?? String(value)
  }

  private static func money(_ amount: Double, symbol: String) -> String {
    let hasFraction = abs(amount.truncatingRemainder(dividingBy: 1)) > 0.0001
    let formatter = NumberFormatter()
    formatter.locale = Locale(identifier: "en_IN")
    formatter.numberStyle = .decimal
    formatter.minimumFractionDigits = hasFraction ? 2 : 0
    formatter.maximumFractionDigits = 2
    let formatted = formatter.string(from: NSNumber(value: abs(amount))) ?? "\(abs(amount))"
    return (amount < 0 ? "−" : "") + symbol + formatted
  }

  static func spent(_ amount: Double, currency: Currency = .INR) -> String {
    "−" + money(amount, currency: currency)
  }

  static func percent(_ value: Double, total: Double) -> Int {
    if total <= 0 { return 0 }
    return Int((value / total * 100).rounded())
  }

  static func compactMoney(_ amount: Double, currency: Currency = .INR) -> String {
    let symbol = currencySymbol(currency)
    if amount >= 1000 {
      let k = String(format: "%.1f", amount / 1000).replacingOccurrences(of: ".0", with: "")
      return "\(symbol)\(k)k"
    }
    let trimmed = String(format: "%.2f", amount)
      .replacingOccurrences(of: #"\.?0+$"#, with: "", options: .regularExpression)
    return "\(symbol)\(trimmed)"
  }
}

enum Greeting {
  static func greetingFor(_ date: Date = Date(), calendar: Calendar = .current) -> String {
    let hour = calendar.component(.hour, from: date)
    if hour < 12 { return "Good morning" }
    if hour < 17 { return "Good afternoon" }
    return "Good evening"
  }
}
